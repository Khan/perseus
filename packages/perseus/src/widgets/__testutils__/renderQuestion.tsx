import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
