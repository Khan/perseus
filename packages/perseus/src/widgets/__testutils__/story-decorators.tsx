import * as React from "react";

export const mobileDecorator = (Story) => (
    <div className="framework-perseus perseus-mobile">
        <Story />
    </div>
);

export const articleDecorator = (Story) => (
    <div className="framework-perseus perseus-article">
        <Story />
    </div>
);

export const mobileArticleDecorator = (Story) => (
    <div className="framework-perseus perseus-mobile perseus-article">
        <Story />
    </div>
);

export const rtlDecorator = (Story) => (
    <div style={{direction: "rtl"}}>
        <Story />
    </div>
);
