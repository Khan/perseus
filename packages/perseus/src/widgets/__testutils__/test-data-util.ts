import type {PerseusItem, PerseusRenderer} from "@khanacademy/perseus-core";

/**
 * Creates a Perseus item with the provided question/PerseusRenderer data.
 * Useful for converting test data used with RendererWithDebugUI to test data
 * for ServerItemRendererWithDebugUI.
 *
 * @param {PerseusRenderer} question - The question to be included in the Perseus item.
 * @returns {PerseusItem} The created Perseus item.
 */
export const createPerseusItem = (question: PerseusRenderer): PerseusItem => {
    return {
        question: question,
        answer: null,
        itemDataVersion: {
            major: 0,
            minor: 1,
        },
        hints: [],
        answerArea: {
            calculator: false,
            chi2Table: false,
            financialCalculatorMonthlyPayment: false,
            financialCalculatorTotalAmount: false,
            financialCalculatorTimeToPayOff: false,
            periodicTable: false,
            periodicTableWithKey: false,
            tTable: false,
            zTable: false,
        },
    };
};
