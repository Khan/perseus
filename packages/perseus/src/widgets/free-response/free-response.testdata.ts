import type {PerseusRenderer} from "@khanacademy/perseus-core";

interface Options {
    allowUnlimitedCharacters?: boolean;
    characterLimit?: number;
}

export function getFreeResponseItemData(options?: Options): PerseusRenderer {
    return {
        content: "[[\u2603 free-response 1]]\n",
        images: {},
        widgets: {
            "free-response 1": {
                graded: false,
                version: {major: 0, minor: 0},
                static: false,
                type: "free-response",
                options: {
                    allowUnlimitedCharacters:
                        options?.allowUnlimitedCharacters ?? false,
                    characterLimit: options?.characterLimit ?? 500,
                    placeholder: "test-placeholder",
                    question: "test-question",
                    scoringCriteria: [
                        {
                            text: "test-criterion",
                        },
                    ],
                },
                alignment: "default",
            },
        },
    };
}
