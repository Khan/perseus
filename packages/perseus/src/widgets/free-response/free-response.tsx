/**
 * This widget is used for freeform text input. It is configured with a question
 * as an option and renders the question text along with a text input area where
 * the user can type any text as their answer. The initial use case for this widget
 * is "short answer" type questions.
 */

import {Text, View} from "@khanacademy/wonder-blocks-core";
import {TextArea} from "@khanacademy/wonder-blocks-form";
import {PhosphorIcon} from "@khanacademy/wonder-blocks-icon";
import {LabeledField} from "@khanacademy/wonder-blocks-labeled-field";
import {
    color,
    font,
    spacing,
    semanticColor,
} from "@khanacademy/wonder-blocks-tokens";
import warningCircleIcon from "@phosphor-icons/core/regular/warning-circle.svg";
import {StyleSheet} from "aphrodite";
import * as React from "react";

import {PerseusI18nContext} from "../../components/i18n-context";
import Renderer from "../../renderer";

import type {Widget, WidgetExports, WidgetProps} from "../../types";
import type {
    PerseusFreeResponseUserInput,
    PerseusFreeResponseWidgetOptions,
} from "@khanacademy/perseus-core";

type Props = WidgetProps<
    PerseusFreeResponseWidgetOptions,
    PerseusFreeResponseUserInput
>;

type DefaultProps = Pick<Props, "userInput">;

// TODO(agoforth): Create a custom validator for the widget that will cause
//   renderer.emptyWidgets() to work when there is no user input.

export class FreeResponse extends React.Component<Props> implements Widget {
    static contextType = PerseusI18nContext;
    declare context: React.ContextType<typeof PerseusI18nContext>;

    // this just helps with TS weak typing when a Widget
    // doesn't implement any Widget methods
    isWidget = true as const;

    static defaultProps: DefaultProps = {
        userInput: {
            currentValue: "",
        },
    };

    characterCount = () => {
        return this.props.userInput.currentValue.replace(/\n/g, "").length;
    };

    _handleUserInput = (newValue: string) => {
        this.props.handleUserInput({currentValue: newValue});
    };

    isOverLimit() {
        return (
            !this.props.allowUnlimitedCharacters &&
            this.characterCount() > this.props.characterLimit
        );
    }

    renderCharacterCount(): React.ReactNode {
        if (this.props.allowUnlimitedCharacters) {
            return null;
        }

        const characterCountText = this.context.strings.characterCount({
            used: this.characterCount(),
            num: this.props.characterLimit,
        });

        return (
            <View>
                <Text
                    role="status"
                    style={[
                        styles.characterCountText,
                        this.isOverLimit()
                            ? styles.overCharacterLimit
                            : undefined,
                    ]}
                >
                    {this.isOverLimit() && (
                        <PhosphorIcon
                            icon={warningCircleIcon}
                            size="small"
                            style={styles.warningCircleIcon}
                        />
                    )}

                    {characterCountText}
                </Text>
            </View>
        );
    }

    render(): React.ReactNode {
        return (
            <View style={styles.container} className={"free-response"}>
                <LabeledField
                    label={
                        <View className="free-response-question">
                            <Renderer
                                content={this.props.question}
                                strings={this.context.strings}
                            />
                        </View>
                    }
                    field={
                        <TextArea
                            error={this.isOverLimit()}
                            onChange={this._handleUserInput}
                            placeholder={this.props.placeholder}
                            style={styles.textarea}
                            value={this.props.userInput.currentValue}
                        />
                    }
                />
                {this.renderCharacterCount()}
            </View>
        );
    }
}

function getStartUserInput(): PerseusFreeResponseUserInput {
    return {
        currentValue: "",
    };
}

export default {
    name: "free-response",
    accessible: true,
    displayName: "Free Response (Assessments only)",
    widget: FreeResponse,
    hidden: false,
    // FreeResponse doesn't serialize user input,
    // so just bring up the default user input when restoring
    // (which we likely never should/will for FreeResponse)
    getUserInputFromSerializedState: getStartUserInput,
    getStartUserInput,
} as WidgetExports<typeof FreeResponse>;

const styles = StyleSheet.create({
    container: {
        gap: spacing.xSmall_8,
    },
    characterCountText: {
        color: semanticColor.core.foreground.neutral.default,
        fontSize: font.size.small,
    },
    overCharacterLimit: {
        color: color.red,
    },
    textarea: {
        padding: spacing.medium_16,
    },
    warningCircleIcon: {
        marginInlineEnd: spacing.xSmall_8,
    },
});
