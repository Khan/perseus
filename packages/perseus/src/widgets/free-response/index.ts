export {default} from "./free-response";
