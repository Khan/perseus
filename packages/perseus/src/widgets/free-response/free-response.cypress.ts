import renderQuestion from "../../../../../testing/render-question-with-cypress";
import {cypressTestDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import * as Perseus from "../../index";

import {question} from "./free-response.testdata";

describe("FreeResponse Widget", () => {
    beforeEach(() => {
        Dependencies.setDependencies(cypressTestDependencies);
        Perseus.init();
    });

    it("allows text input", () => {
        // Arrange
        renderQuestion(question);

        // Act - type some text
        cy.get("textarea").focus().type("my answer");

        // Assert
        cy.focused().should("have.text", "my answer");
    });
});
