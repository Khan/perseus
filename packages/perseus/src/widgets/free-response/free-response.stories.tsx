import {FreeResponse} from "./free-response";

import type {Meta, StoryObj} from "@storybook/react-vite";

const meta: Meta<typeof FreeResponse> = {
    component: FreeResponse,
    title: "Widgets/Free Response",
    tags: ["!dev"],
    parameters: {
        docs: {
            description: {
                component:
                    "A widget that provides a text area for users to enter open-ended responses,\
                    supporting long-form answers and essay-type questions.",
            },
        },
    },
};

export default meta;
type Story = StoryObj<typeof FreeResponse>;

export const Primary: Story = {
    args: {
        allowUnlimitedCharacters: false,
        characterLimit: 500,
        placeholder: "Enter your answer here",
        question: "What is the theme of the essay?",
    },
};

export const CharacterLimit: Story = {
    args: {
        allowUnlimitedCharacters: false,
        characterLimit: 500,
        placeholder: "Enter your answer here",
        question: "What is the theme of the essay?",
    },
};

export const BoldedQuestion: Story = {
    args: {
        allowUnlimitedCharacters: false,
        characterLimit: 500,
        placeholder: "Enter your answer here",
        question: "**What is the theme of the essay?**",
    },
};

export const UnlimitedCharacters: Story = {
    args: {
        allowUnlimitedCharacters: true,
        characterLimit: 500,
        placeholder: "Enter your answer here",
        question:
            "What is the theme of the essay?\n\n**Put your answer in your own words.**",
    },
};

export const QuestionWithTex: Story = {
    name: "Question with TeX content",
    args: {
        allowUnlimitedCharacters: true,
        characterLimit: 500,
        placeholder: "Enter your answer here",
        question:
            "What changes are required to solve the following equation? $\\dfrac{6-3}{1-0}=\\dfrac{3}{1}=3$",
    },
};
