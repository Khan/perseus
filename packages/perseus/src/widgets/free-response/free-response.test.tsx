import {screen} from "@testing-library/react";
import {userEvent as userEventLib} from "@testing-library/user-event";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {renderQuestion} from "../__testutils__/renderQuestion";

import {getFreeResponseItemData} from "./free-response.testdata";

import type {APIOptions} from "../../types";
import type {UserEvent} from "@testing-library/user-event";

describe("free-response widget", () => {
    let userEvent: UserEvent;

    beforeEach(() => {
        userEvent = userEventLib.setup({
            advanceTimers: jest.advanceTimersByTime,
        });

        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should snapshot", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };

        // Act
        const {container} = renderQuestion(
            getFreeResponseItemData(),
            apiOptions,
        );

        // Assert
        expect(container).toMatchSnapshot("first render");
    });

    it("should snapshot on mobile", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: true,
        };

        // Act
        const {container} = renderQuestion(
            getFreeResponseItemData(),
            apiOptions,
        );

        // Assert
        expect(container).toMatchSnapshot("first mobile render");
    });

    it("should render the character limit when not allowing unlimited characters", () => {
        // Arrange
        const data = getFreeResponseItemData({allowUnlimitedCharacters: false});

        // Act
        renderQuestion(data, {});

        // Assert
        expect(screen.queryByText(/Character/)).toBeVisible();
    });

    it("should not render the character limit when allowing unlimited characters", () => {
        // Arrange
        const data = getFreeResponseItemData({allowUnlimitedCharacters: true});

        // Act
        renderQuestion(data, {});

        // Assert
        expect(screen.queryByText(/Character/)).not.toBeInTheDocument();
    });

    it("should ignore newline characters when rendering the character limit", async () => {
        // Arrange
        const data = getFreeResponseItemData({characterLimit: 10});

        // Act
        renderQuestion(data, {});

        await userEvent.type(
            screen.getByRole("textbox", {name: "test-question"}),
            "a\nb",
        );

        // Assert
        expect(screen.queryByText(/2 \/ 10 Character/)).toBeVisible();
    });

    it("should render the question text", () => {
        // Arrange
        // Act
        renderQuestion(getFreeResponseItemData(), {});

        // Assert
        expect(
            screen.getByRole("textbox", {name: "test-question"}),
        ).toBeVisible();
    });

    it("should render the placeholder text", () => {
        // Arrange
        // Act
        renderQuestion(getFreeResponseItemData(), {});

        // Assert
        expect(
            screen.getByRole("textbox", {name: "test-question"}),
        ).toHaveAttribute("placeholder", "test-placeholder");
    });

    it("should return the correct user input", async () => {
        // Arrange
        const {renderer} = renderQuestion(getFreeResponseItemData());
        await userEvent.type(
            screen.getByLabelText("test-question"),
            "test-answer",
        );

        // Act
        const userInput = renderer.getUserInputMap();

        // Assert
        expect(userInput).toMatchObject({
            "free-response 1": {
                currentValue: "test-answer",
            },
        });
    });
});
