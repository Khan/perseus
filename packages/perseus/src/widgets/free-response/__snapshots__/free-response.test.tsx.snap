// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`free-response widget should snapshot on mobile: first mobile render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="default_xu2jcg-o_O-container_lci7r free-response"
          >
            <div
              class="default_xu2jcg"
            >
              <div
                class="default_xu2jcg-o_O-labelContainer_1qlbwdv"
              >
                <label
                  class="text_f1191h-o_O-BodyText_1xtvlx3-o_O-BodyTextMediumSemiWeight_w58hss-o_O-label_1r6q8ju-o_O-labelWithNoDescription_i032sc"
                  for=":r2:-labeled-field-field"
                  id=":r2:-labeled-field-label"
                >
                  <div
                    class="default_xu2jcg free-response-question"
                  >
                    <div
                      class="perseus-renderer perseus-renderer-responsive"
                    >
                      <div
                        class="paragraph"
                        data-perseus-paragraph-index="0"
                      >
                        <div
                          class="paragraph"
                        >
                          test-question
                        </div>
                      </div>
                    </div>
                  </div>
                </label>
              </div>
              <div
                class="default_xu2jcg-o_O-inlineStyles_zdxht7"
              >
                <textarea
                  aria-describedby=""
                  aria-invalid="false"
                  aria-required="false"
                  class="textarea_xrj3rm-o_O-BodyTextMediumMediumWeight_ijclyr-o_O-default_8fhii8-o_O-textarea_1x0fg6n"
                  id=":r2:-labeled-field-field"
                  placeholder="test-placeholder"
                />
              </div>
              <div
                aria-atomic="true"
                aria-live="assertive"
                class="default_xu2jcg-o_O-helperTextWithIcon_1hmzxfl"
                id=":r2:-labeled-field-error"
              />
            </div>
            <div
              class="default_xu2jcg"
            >
              <span
                class="text_f1191h-o_O-characterCountText_1iy49wk"
                role="status"
              >
                0 / 500 Characters
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`free-response widget should snapshot: first render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="default_xu2jcg-o_O-container_lci7r free-response"
          >
            <div
              class="default_xu2jcg"
            >
              <div
                class="default_xu2jcg-o_O-labelContainer_1qlbwdv"
              >
                <label
                  class="text_f1191h-o_O-BodyText_1xtvlx3-o_O-BodyTextMediumSemiWeight_w58hss-o_O-label_1r6q8ju-o_O-labelWithNoDescription_i032sc"
                  for=":r0:-labeled-field-field"
                  id=":r0:-labeled-field-label"
                >
                  <div
                    class="default_xu2jcg free-response-question"
                  >
                    <div
                      class="perseus-renderer perseus-renderer-responsive"
                    >
                      <div
                        class="paragraph"
                        data-perseus-paragraph-index="0"
                      >
                        <div
                          class="paragraph"
                        >
                          test-question
                        </div>
                      </div>
                    </div>
                  </div>
                </label>
              </div>
              <div
                class="default_xu2jcg-o_O-inlineStyles_zdxht7"
              >
                <textarea
                  aria-describedby=""
                  aria-invalid="false"
                  aria-required="false"
                  class="textarea_xrj3rm-o_O-BodyTextMediumMediumWeight_ijclyr-o_O-default_8fhii8-o_O-textarea_1x0fg6n"
                  id=":r0:-labeled-field-field"
                  placeholder="test-placeholder"
                />
              </div>
              <div
                aria-atomic="true"
                aria-live="assertive"
                class="default_xu2jcg-o_O-helperTextWithIcon_1hmzxfl"
                id=":r0:-labeled-field-error"
              />
            </div>
            <div
              class="default_xu2jcg"
            >
              <span
                class="text_f1191h-o_O-characterCountText_1iy49wk"
                role="status"
              >
                0 / 500 Characters
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
